crates/l2/sequencer/l1_proof_sender.rs

```diff
@@ -0,0 +1,248 @@
+            proof_coordinator::ProofCoordinatorConfig,
+        },
+        prover::{
+            proving_systems::ProverType,
+            save_state::{block_number_has_all_needed_proofs, read_proof, StateFileType},
+        },
+    },
+};
+
+// These constants have to match with the OnChainProposer.sol contract
+const R0VERIFIER: &str = "R0VERIFIER()";
+const SP1VERIFIER: &str = "SP1VERIFIER()";
+const PICOVERIFIER: &str = "PICOVERIFIER()";
+const VERIFIER_CONTRACTS: [&str; 3] = [R0VERIFIER, SP1VERIFIER, PICOVERIFIER];
+const DEV_MODE_ADDRESS: H160 = H160([
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0xAA,
+]);
+const VERIFY_FUNCTION_SIGNATURE: &str =
+    "verify(uint256,bytes,bytes32,bytes32,bytes32,bytes,bytes,bytes32,bytes,uint256[8])";
+
+pub async fn start_l1_proof_sender() -> Result<(), ConfigError> {
+    let eth_config = EthConfig::from_env()?;
+    let committer_config = CommitterConfig::from_env()?;
+    let prover_server_config = ProofCoordinatorConfig::from_env()?;
+
+    let proof_sender =
+        L1ProofSender::new(&prover_server_config, &committer_config, &eth_config).await?;
+    proof_sender.run().await;
+
+    Ok(())
+}
+
+struct L1ProofSender {
+    eth_client: EthClient,
+    l1_address: Address,
+    l1_private_key: SecretKey,
+    on_chain_proposer_address: Address,
+    needed_proof_types: Vec<ProverType>,
+}
+
+impl L1ProofSender {
+    async fn new(
+        config: &ProofCoordinatorConfig,
+        committer_config: &CommitterConfig,
+        eth_config: &EthConfig,
+    ) -> Result<Self, ConfigError> {
+        let eth_client = EthClient::new(&eth_config.rpc_url);
+
+        let verifier_contracts = EthClient::get_verifier_contracts(
+            &eth_client,
+            &VERIFIER_CONTRACTS,
+            committer_config.on_chain_proposer_address,
+        )
+        .await?;
+
+        let mut needed_proof_types = vec![];
+        if !config.dev_mode {
+            for (key, addr) in verifier_contracts {
+                if addr == DEV_MODE_ADDRESS {
+                    continue;
+                } else {
+                    match key.as_str() {
+                        "R0VERIFIER()" => {
+                            info!("RISC0 proof needed");
+                            needed_proof_types.push(ProverType::RISC0);
+                        }
+                        "SP1VERIFIER()" => {
+                            info!("SP1 proof needed");
+                            needed_proof_types.push(ProverType::SP1);
+                        }
+                        "PICOVERIFIER()" => {
+                            info!("PICO proof needed");
+                            needed_proof_types.push(ProverType::Pico);
+                        }
+                        _ => unreachable!("There shouldn't be a value different than the used backends/verifiers R0VERIFIER|SP1VERIFER|PICOVERIFIER."),
```

Comment:
Maybe use an enum for the verifiers (makes sense since it'd be extended in time), or return an internal error instead of using `unreachable` (if the internal error is correctly handled).

Reply:
Done [0cb451c](https://github.com/lambdaclass/ethrex/pull/2478/commits/0cb451c6c8958b577e0b79c90a613a1f4ec132a2).